### ISSUE TITLE
user.buyeruid is not being added to the request sent to the Bidder

### ISSUE THREAD

**Issue by snigel-nelson** [prebid/prebid-server#2238]
## Description of the Issue

The issue I'm reporting is that the _user.buyeruid_ attribute is not being passed in the request sent to **Rubicon**.

 ### Context
I'm using a custom version of prebid-server (PBS) to allow publishers to have **S2S HB**  if they want to take advantage of it. In order to get better bids, and to fulfill certain mandatory requirements of some bidders, the _buyeruid_ attribute must be present in the sent request. It turns out that this is not the case for Rubicon.
On the client-side I am using **v5.20.2** of **Prebid.js** and its _prebidServerBidAdapter_.
I'm also using Stored Requests (_ext.prebid.storedrequest_) and Stored Impressions (_ext.imp[n].ext.prebid.storedrequest_) instead of sending the request with all the bidders and attributes needed.
On the server-side my PBS version is **v0.205.0**.

## Tests
User ID sync endpoint is working correctly. The endpoint defined by Rubicon is manually configured depending on the region of the order (EU, US, Asia,...).
The _buyeruid_ is successfully written in the PBS' cookie and using the _pbjs_debug=true_ flag I can analyze the ext.debug.httpcalls of each bidder and check if the _buyeruid_ is sent in the request to all bidders (OpenX, IX, AppNexus,... ) except **Rubicon**.
I can assert that these _buyeruids_ are the same contained in the cookie by using the **/getuids** endpoint (https://prebid-server.com/getuids).

## Cause of the issue
After performing a battery of tests I was able to identify the indirect reason why the _buyeruids_ is not being sent in the request to Rubicon. It has something to do with the adapter and how it handles with "keywords" present in the request from the client.
If I add the _ext.prebid.bidderconfig_ object to the request in order to be able to define keywords, Rubicon's prebid-server adapter doesn't pass the _buyeruid_ to the bidder. This happens for the release 0.205.0. For the release 0.179.0 no problem is detected, I can add this object and the _buyeruid_ is correctly passed.
I'll share with you an example of the request from the client to my prebid-server:
`
{
  (...)
  "ext": {
        "prebid": {
            "auctiontimestamp": 1651575956267,
            "targeting": {
                "includewinners": true,
                "includebidderkeys": false
            },
            "channel": {
                "name": "pbjs",
                "version": "v5.20.2"
            },
            "debug": true,
            "storedrequest": {
                "id": "stored_req_1"
            },
            "bidderconfig": [{
                    "bidders": ["bidder"],
                    "config": {
                        "ortb2": {
                            "site": {
                                "ext": {
                                    "data": {
                                        "keywords": "keyword1"
                                    }
                                }
                            }
                        }
                    }
                }
            ]
        }
    },
  (...)
}
`

Here I only show the part of the order that really matters. Removing this object the same request works also for PBS version 0.205.0
For version 0.179.0 it works fine with _ext.prebid.bidderconfig_.

## Expected behavior ##

The result I expected is the something like this:
`
{
"id": "13d42e0c-2d9c-4668-a5ea-abb3d6baa8b2",
    "imp": [...],
    (...),
     "user": {
        "buyeruid": "L2OO88JF-X-83MF",
        "ext": {
            "consent": "abcdefgh",
            "eids": [{
                    "source": "source_1",
                    "uids": [{
                            "id": "id_1",
                            "atype": 1
                        }
                    ]
                }, {
                    "source": "source_2",
                    "uids": [{
                            "id": "id_2",
                            "atype": 1,
                            "ext": {
                                "linkType": 1
                            }
                        }
                    ]
                }
            ],
            "rp": {
                "target": {}
            }
        }
    },
  (...)
}
`

## Questions

- Is this a problem with the prebid-server v0.205.0 and its Rubicon's adapter? Something changed in this adapter between v0.179.0 and v0.205.0
- How exactly the _user.buyeruid_ is inserted in the final request sent to the Bidder?

I'll be looking forward for your reply.

As usual, if you need more info just let me know.

Thanks,

Nelson

**Comment by bsardo** [comment]
Hi @snigel-nelson,

A quick diff between v0.179.0 and v0.205.0 turns up three PRs in v0.205.0 that aren't in v0.179.0 that might be of interest here:
1. #1906 - First Party Data
2. #2091 - Rubicon: update fpd fields resolving
3. #2138  - Rubicon: Add one more way for Rubicon bidder to represent User.BuyerUid

First off, it appears you're specifying `keywords` as first party data given your use of `ext.prebid.bidderconfig`. The first party data feature was new to PBS-Go between v0.179.0 and v.205.0, though I think the Rubicon adapter did offer some FPD support. `ext.prebid.bidderconfig` was most likely being ignored in v0.179.0.

In order to assist you here, I think I need some more information.

Can you provide the following:
- the effective resolved request (your request after it has been merged with any applicable stored requests/imps)
- your config file

Thanks,
Brian

**Comment by snigel-nelson** [comment]
Hi @bsardo,

Thanks for your reply and apologies for not answering sooner. In fact, there were some changes between v0.179.0 and v0.205.0. At the moment I've checked and haven't found any specific changes that could affect the inclusion of the _buyeruids_ attribute in the request sent to Rubicon. I share the same opinion, that in version 0.179.0 the _ext.prebid.bidderconfig_ object would be ignored. Anyway, I'll double-check the changes made.

I will then share with you the data you requested in the attached files:

-  resolved request. The full request after all Stored Imps and Stored Request are merged. 
[resolved_request.txt](https://github.com/prebid/prebid-server/files/8678723/resolved_request.txt)
- Rubicon's request body (JSON format). The request was sent from our PBS to Rubicon. 
[rubicon_requestBody.txt](https://github.com/prebid/prebid-server/files/8678714/rubicon_requestBody.txt)
- a cookie with all the _uids_. 
[cookie.txt](https://github.com/prebid/prebid-server/files/8678704/cookie.txt)
- _pbs.yaml_ file from our PBS (config file). 
[pbs.txt](https://github.com/prebid/prebid-server/files/8678719/pbs.txt)


Note: For privacy reasons some attributes' values were changed.

I hope this info is useful to you and you catch the problem. If you need anything else just let me know.

Thanks, 
Nelson

**Comment by snigel-nelson** [comment]
> Hi @snigel-nelson,
> 
> A quick diff between v0.179.0 and v0.205.0 turns up three PRs in v0.205.0 that aren't in v0.179.0 that might be of interest here:
> 
> 1. [First party data #1906](https://github.com/prebid/prebid-server/pull/1906)
> 2. [Rubicon: update fpd fields resolving #2091](https://github.com/prebid/prebid-server/pull/2091)
> 3. [Rubicon: Add one more way for Rubicon bidder to represent `User.BuyerUid`. #2138](https://github.com/prebid/prebid-server/pull/2138)
> 
> First off, it appears you're specifying `keywords` as first party data given your use of `ext.prebid.bidderconfig`. The first party data feature was new to PBS-Go between v0.179.0 and v.205.0, though I think the Rubicon adapter did offer some FPD support. `ext.prebid.bidderconfig` was most likely being ignored in v0.179.0.
> 
> In order to assist you here, I think I need some more information.
> 
> Can you provide the following:
> 
> * the effective resolved request (your request after it has been merged with any applicable stored requests/imps)
> * your config file
> * your cookie
> 
> Thanks, Brian

Hi @bsardo,

I hope everything is fine with you. Do you have any news regarding this issue. Do you need more info provided besides the one I shared with you in my previous comment?

Thanks,
Nelson

**Comment by bsardo** [comment]
Hi @snigel-nelson,
Sorry for the slow response and thank you for providing all of the requested information. I'm working on your issue at the moment. I was able to reproduce the problem so I don't think I need any additional information at this point in time. I'll be in touch shortly.
Brian

**Comment by bsardo** [comment]
Hi @snigel-nelson,

I've identified the root of the problem. The issue is not specific to rubicon, as I was able to replicate it for other adapters, but rather there is a bug in our first party data logic.

First party data currently works like this:
1) In `exchange/exchange.go#HoldAuction`, we call `ExtractFPDForBidders` which checks if there is any FPD defined for each bidder. If app, site and/or user FPD is defined for a bidder, the final app, site and/or user object is constructed from the base request, stored request and the applicable FPD for that bidder.
2) After each bid request has been created, in `exchange/utils.go#cleanOpenRTBRequests.go, `applyFPD` is called for each bidder that had FPD. It is here that any app/site/user objects constructed in step 1 are applied to the bid request overwriting the object in its entirety. Note that if no app/site/user object was constructed in step 1 for a given bidder, then the app/site/user defined in the base of the original request would be used.

In between these two steps the cookie is analyzed and the buyeruid is set on the original request in `user.buyeruid` if the ID exists in the cookie but was not in the original request.

Now this doesn't seem like it should be an issue in your example because you didn't specify any `user` FPD, as far as I can tell, from the data you provided, just `site` FPD. As such, I wouldn't have expected `ExtractFPDForBidders` to construct a new `user` object to apply to the rubicon bid request. However, it **is** creating a new `user` object (which doesn't contain the buyeruid) and that is being applied to the rubicon bidder which overwrites the original request `user` object that was mutated to include the buyeruid from the cookie. So this is the first problem we need to fix to address your scenario. If there isn't any `user` FPD, we shouldn't be creating a new `user` object; it should be `nil`.

On that same note, I suspect that there is another issue here. If you **had** provided `user` FPD I think we still would have lost the buyeruid because the constructed `user` object with FPD would have overwritten the original `user` object that was mutated to include the buyeruid when the cookie was read. We need to make an adjustment to ensure that any non-FPD mutations to the original request (in this case grabbing the buyeruid from the cookie and applying it) occur before `ExtractFPDForBidders` is called so the data isn't potentially overwritten when applying the constructed `user` object with FPD to a bidder request.

I'll talk to the team and see if we can get started on this after the holiday.

Brian

**Comment by snigel-nelson** [comment]
Hi @bsardo,

Despite having signaled with the emojis I want to thank you again for the detailed reply you wrote, explaining to me what's causing this issue, the FPD logic. How FPD is implemented and working, and how can affect this specific case (I don't provide any `user` FDP) or other situations (if you provide an `user` FPD). Most useful.

Since it has passed a while since the last comment, do you have any news regarding this subject?

Thanks,
Nelson

**Comment by bsardo** [comment]
Hi @snigel-nelson,

We're planning to start work on this in the next couple of days once we close out the current open FPD PR. @VeronikaSolovei9 will be handling this one.

Brian

**Comment by VeronikaSolovei9** [comment]
@bsardo 
Brian, thank you for your investigation! It helped a lot!  Here is what I found:
`So this is the first problem we need to fix to address your scenario. If there isn't any user FPD, we shouldn't be creating a new user object in ExtractFPDForBidders; it should be nil.` - user is present in request, hence we need to have in FPD to modify it if needed and to apply it later. There are just one conditions where we don't create user in FPD (User is not present in bidder config fpd and not present in request):
```
	if bidRequestUser == nil && fpdConfigUser == nil {
		return nil, nil
	}
```
This is why we have  user for Rubicon in this case. Then using this approach we know exactly what to apply for final bidder request: 
```
func applyFPD(fpd *firstpartydata.ResolvedFirstPartyData, bidReq *openrtb2.BidRequest) {
	if fpd.Site != nil {
		bidReq.Site = fpd.Site
	}
	if fpd.App != nil {
		bidReq.App = fpd.App
	}
	if fpd.User != nil {
		bidReq.User = fpd.User
	}
}
In case there is no FPD config user - user remains unchanged.

```

`If you had provided user FPD I think we still would have lost the buyeruid because the constructed user object with FPD would have overwritten the original user object that was mutated to include the buyeruid when the cookie was read` - you are right about it.
Fix PR: https://github.com/prebid/prebid-server/pull/2269

**Comment by mansinahar** [comment]
Closing this issue out as the PR is now merged. Please reopen if needed. Thanks!

**Comment by snigel-nelson** [comment]
Hi @mansinahar, 
Ya, you're correct, the PR with the solution was already merged and a release launched (v0.214.0), so no need to keep it open while I run my tests, like I was thinking of. 
Thanks to all the team: @bsardo , @VeronikaSolovei9 , @bretg , @mansinahar and others.
